pkg/kubelet/gpu/nvidia/nvidia_gpu_manager.go

```diff
@@ -202,13 +208,44 @@ func (ngm *nvidiaGPUManager) discoverGPUs() error {
 		}
 		if reg.MatchString(f.Name()) {
 			glog.V(2).Infof("Found Nvidia GPU %q", f.Name())
-			ngm.allGPUs.Insert(path.Join(devDirectory, f.Name()))
+			foundGPUs.Insert(f.Name())
 		}
 	}
-
+	if foundGPUs.Len() != expected {
+		err := fmt.Errorf("expected %d GPUs, found %d", expected, foundGPUs.Len())
+		return err 
+	}
+	for _, gpuToAdd := range foundGPUs.List() {
+		ngm.allGPUs.Insert(path.Join(devDirectory, gpuToAdd))
+		
+	}		
 	return nil
 }
 
+// expectedGPUs reads the devices procfile and determines how many GPUs we should find
+// when discoverGPUs is called
+// TODO: This most likely will count dead GPUs and should be refactored if we
```

Comment:
Linked in my NVML proposal for now. I can split that proposal up to just suit the general use case of discovery/health-checking.